test/hotspot/jtreg/runtime/ClassInitErrors/InitExceptionUnloadTest.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/**
+ * @test
+ * @bug 8048190
+ * @summary Test that the NCDFE saves original exception during class initialization,
+ *          and doesn't prevent the classes in the stacktrace to be unloaded.
+ * @requires vm.opt.final.ClassUnloading
+ * @modules java.base/jdk.internal.misc
+ * @library /test/lib
+ * @build sun.hotspot.WhiteBox
+ * @run driver jdk.test.lib.helpers.ClassFileInstaller sun.hotspot.WhiteBox
+ * @run main/othervm -Xbootclasspath/a:. -Xmn8m -XX:+UnlockDiagnosticVMOptions -Xlog:class+unload -XX:+WhiteBoxAPI InitExceptionUnloadTest
+ */
+
+import java.io.ByteArrayOutputStream;
+import java.io.PrintStream;
+
+import sun.hotspot.WhiteBox;
+import jdk.test.lib.classloader.ClassUnloadCommon;
+
+public class InitExceptionUnloadTest {
+    static public class ThrowsRuntimeException { static int x = 1/0; }
+    static public class ThrowsError { static { if (true) throw new Error(); } }
+    static public class SpecialException extends RuntimeException {
+        SpecialException(int count, String message) {
+            super(message + count);
+        }
+    }
+    static public class ThrowsSpecialException {
+        static {
+            if (true) foo();
```

Comment:
Not sure why you need to define foo() - and if you use foo() you shouldn't need the "if (true)".

Reply:
I've removed the foos.  They made jasm hacking easier when I thought I needed that.